File: CppSpeedTest - Copy/NeuronArrayBase.cpp
#include "NeuronArrayBase.h"
#include <windows.h>
#include <ppl.h>
#include <iostream>
#include <random>

using namespace concurrency;
using namespace std;

namespace NeuronEngine
{
	NeuronArrayBase::NeuronArrayBase()
	{
	}
	NeuronArrayBase::~NeuronArrayBase()
	{
	}

	void NeuronArrayBase::Initialize(int theSize, NeuronBase::modelType t)
	{
		arraySize = theSize;
		neuronArray.reserve(arraySize);
		for (int i = 0; i < arraySize; i++)
		{
			NeuronBase n(i);
			neuronArray.push_back(n);
		}
	}
	long NeuronArrayBase::GetGeneration()
	{
		return generation;
	}
	NeuronBase* NeuronArrayBase::GetNeuron(int i)
	{
		return &neuronArray[i];
	}


	int NeuronArrayBase::GetArraySize()
	{
		return arraySize;
	}

	long NeuronArrayBase::GetTotalSynapseCount()
	{
		long count = 0;
		for (int i = 0; i < arraySize; i++)
			if (GetNeuron(i)->GetSynapses() != NULL)
			{
				count += (unsigned)GetNeuron(i)->GetSynapses()->size();
			}
		return count;;
	}

	void NeuronArrayBase::GetBounds1(int taskID, int& start, int& end)
	{
		int numberToProcess = arraySize / threadCount;
		int remainder = arraySize % threadCount;
		start = numberToProcess * taskID;
		end = start + numberToProcess;
		if (taskID < remainder)
		{
			start += taskID;
			end = start + numberToProcess + 1;
		}
		else
		{
			start += remainder;
			end += remainder;
		}
	}
	void NeuronArrayBase::Fire()
	{
		generation++;
		firedCount = 0;
		parallel_for(0, threadCount, [&](int value) {
			ProcessNeurons1(value);
			});
		parallel_for(0, threadCount, [&](int value) {
			ProcessNeurons2(value);
			});
	}
	int NeuronArrayBase::GetFiredCount()
	{
		return firedCount;
	}
	int NeuronArrayBase::GetThreadCount()
	{
		return threadCount;
	}
	void NeuronArrayBase::SetThreadCount(int i)
	{
		threadCount = i;
	}
	void NeuronArrayBase::ProcessNeurons1(int taskID)
	{
		int start, end;
		GetBounds1(taskID, start, end);
		for (int i = start; i < end; i++)
			if (GetNeuron(i)->Fire2NoQ(generation))
				firedCount++;
	}
	void NeuronArrayBase::ProcessNeurons2(int taskID)
	{
		int start, end;
		GetBounds1(taskID, start, end);
		for (int i = start; i < end; i++)
			GetNeuron(i)->Fire1NoQ();
	}


	//firingQueue.clear();
	//firingQueue.insert(firingQueue.end(), manualFire.begin(), manualFire.end());
	//manualFire.clear();
	//for (int i = 0; i < taskCount; i++)
	//{
	//	firingQueue.insert(firingQueue.end(), nextQueue[i].begin(), nextQueue[i].end());
	//	nextQueue[i].clear();
	//	firingQueue.insert(firingQueue.end(), firedQueue[i].begin(), firedQueue[i].end());
	//	firedQueue[i].clear();
	//	firingQueue.insert(firingQueue.end(), zeroQueue[i].begin(), zeroQueue[i].end());
	//	zeroQueue[i].clear();
	//} //spin off the paralel tasks to run the engine on the firing queue
	//firedCount = firingQueue.size();
	//parallel_for(0, taskCount, [&](int value) {
	//	ProcessFiringQueue(value);
	//	});
	//parallel_for(0, taskCount, [&](int value) {
	//	ProcessFiringQueue1(value);
	//	});


		//for (int i = 0; i < taskCount; i++)
		//{
		//	nextQueue.push_back(std::vector<int>());
		//	firedQueue.push_back(std::vector<int>());
		//	zeroQueue.push_back(std::vector<int>());
		//	nextQueue[i].reserve(queuesize / taskCount);
		//	firedQueue[i].reserve(queuesize / taskCount);
		//	zeroQueue[i].reserve(queuesize / taskCount);
		//}
	//void NeuronArrayBase::AddToFiringQueue(int id)
	//{
	//	//manualFire.push_back(id);
	//	NeuronBase* n = GetNeuron(id);
	//	n->currentCharge = 1;
	//	n->lastCharge = 1;
	//}
	//void NeuronArrayBase::ProcessFiringQueue(int id)
	//{ //determine the portion of the firing queue this thread should handle
	//	int taskID = id;
	//	int start, end;
	//	GetBounds(taskID, start, end);
	//	vector<int>& fired = firedQueue[taskID];
	//	//first-phase neuron processing
	//	for (int i = start; i < end; i++)
	//	{
	//		GetNeuron(firingQueue[i])->Fire2(fired);
	//	}
	//}

	//void NeuronArrayBase::ProcessFiringQueue1(int id)
	//{
	//	int taskID = id;
	//	int start, end;
	//	GetBounds(taskID, start, end);
	//	vector<int>& next = nextQueue[taskID];
	//	vector<int>& zero = zeroQueue[taskID];
	//	for (int i = start; i < end; i++)
	//	{
	//		GetNeuron(firingQueue[i])->Fire1(this->neuronArray, next, zero);
	//	}
	//}

	//void NeuronArrayBase::GetBounds(int taskID, int& start, int& end)
	//{
	//	int numberToProcess = firingQueue.size() / taskCount;
	//	int remainder = firingQueue.size() % taskCount;
	//	start = numberToProcess * taskID;
	//	end = start + numberToProcess;
	//	if (taskID < remainder)
	//	{
	//		start += taskID;
	//		end = start + numberToProcess + 1;
	//	}
	//	else
	//	{
	//		start += remainder;
	//		end += remainder;
	//	}
	//}

}

File: NeuronEngine OLD CS/NeuronArrayBase.cs
using System.Collections.Generic;
using System.Threading.Tasks;
namespace BrainSimulator {
    public class NeuronArrayBase{
        public int arraySize = 10000; // ten thousand neurons to start
        public int rows = 100;
        public int Cols { get => arraySize / rows; }
        public NeuronBase[] baseNeuronArray;
        //Variables for the Neuron Firing Queue
        static int queuesize = 10000000;
        //how many parallel tasks and the array to hold them
        const int taskCount = 120;
        List<int> firingQueue = new List<int>(queuesize);
        List<List<int>> nextQueue = new List<List<int>>(taskCount);
        List<List<int>> zeroQueue = new List<List<int>>(taskCount);
        List<List<int>> firedQueue = new List<List<int>>(taskCount);
        List<int> manualFire = new List<int>();

        public long Generation = 0;
        public void Initialize(int theSize, int theRows, NeuronBase.baseModelType t = NeuronBase.baseModelType.Std)
        {
            arraySize = theSize;
            rows = theRows;
            baseNeuronArray = new NeuronBase[arraySize];
            for (int i = 0; i < arraySize; i++)
                baseNeuronArray[i] = new NeuronBase(i, t);
        }
        public int GetInterimNeuronCount() { return arraySize; }
        public NeuronBase GetNeuron(int i) { return baseNeuronArray[i]; }
        public void SetNeuron(int i, NeuronBase n) { baseNeuronArray[i] = n; }
        public int GetArraySize() { return arraySize; }

        public void AddToFiringQueue(int neuronID)
        {
            manualFire.Add(neuronID);
        }
        public int lastFireCount = 0;

        void ProcessFiringQueue(int id)
        {    //determine the portion of the firing queue this thread should handle
            int taskID = id;
            int start, end;
            GetBounds(taskID, out start, out end);
            //first-phase neuron processing
            List<int> queue = firedQueue[taskID];
            for (int i = start; i < end; i++)
            {
                GetNeuron(firingQueue[i]).Fire2(this, queue);
            }
        }
        void ProcessFiringQueue1(int id)
        {
            int taskID = id;
            int start, end;
            GetBounds(taskID, out start, out end);
            List<int> queue1 = nextQueue[taskID];
            List<int> queue2 = zeroQueue[taskID];
            for (int i = start; i < end; i++)
            {
                GetNeuron(firingQueue[i]).Fire1(this, queue1, queue2);
            }
        }
        private void GetBounds(int taskID, out int start, out int end)
        {
            int numberToProcess = firingQueue.Count / taskCount;
            int remainder = firingQueue.Count % taskCount;
            start = numberToProcess * taskID;
            end = start + numberToProcess;
            if (taskID < remainder)
            {
                start += taskID;
                end = start + numberToProcess + 1;
            }
            else
            {
                start += remainder;
                end += remainder;
            }
        }
        public void Fire()
        {
            Generation++;
            if (nextQueue.Count == 0)
            {
                for (int i = 0; i < taskCount; i++)
                {
                    nextQueue.Add(new List<int>(queuesize / taskCount));
                    firedQueue.Add(new List<int>(queuesize / taskCount));
                    zeroQueue.Add(new List<int>(queuesize / taskCount));
                }
            }
            firingQueue.Clear();
            firingQueue.AddRange(manualFire);
            manualFire.Clear();
            for (int i = 0; i < taskCount; i++)
            {
                firingQueue.AddRange(nextQueue[i]);
                nextQueue[i].Clear();
                firingQueue.AddRange(firedQueue[i]);
                firedQueue[i].Clear();
                firingQueue.AddRange(zeroQueue[i]);
                zeroQueue[i].Clear();
            }//spin off the paralel tasks to run the engine on the firing queue
            ParallelOptions x = new ParallelOptions { MaxDegreeOfParallelism = 128 };
            Parallel.For(0, taskCount, x, i => ProcessFiringQueue(i));
            Parallel.For(0, taskCount, x, i => ProcessFiringQueue1(i));
        }
    }
}
